Media transformation class and different transformations based on subsets of channels

wd60622 opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

## Description
<!--- Describe your changes in detail -->

Allow for media transformation on subset of channels

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [x] Closes #712 
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
<!--- Please list the modules that are affected by this PR by typing an `x` in the boxes below: -->
- [x] MMM
- [ ] CLV
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--968.org.readthedocs.build/en/968/

<!-- readthedocs-preview pymc-marketing end -->

juanitorduz commented on pymc_marketing/mmm/__init__.py:
```diff
@@ -39,13 +39,25 @@
 from pymc_marketing.mmm.delayed_saturated_mmm import MMM, DelayedSaturatedMMM
 from pymc_marketing.mmm.fourier import MonthlyFourier, YearlyFourier
+from pymc_marketing.mmm.media_transformation import (
+    MediaConfig,
+    MediaConfigs,
```
Shall we have more different names? Just as "s" at the end is a typo-rick hehe. For example `MediaConfig` and `MediaConfigList`?

wd60622 replied:
Good idea 😆

juanitorduz commented on pymc_marketing/mmm/media_transformation.py:
```diff
@@ -0,0 +1,278 @@
+    ]
+
+
+Apply the media transformation to media data in PyMC model:
```
Can this configuration specification be used through the `media_config` parameter with the `MMM` model class? If so, could you provide an example here as well 🙏 ?

wd60622 replied:
There isn't a `media_config` parameter at the moment. But we could make one 👀

juanitorduz reviewed (commented):
Wow! Super Cool! Just left some questions.

wd60622 commented on pymc_marketing/mmm/media_transformation.py:
```diff
@@ -0,0 +1,278 @@
+    media_transformation: MediaTransformation
+
+
+MediaConfigs = list[MediaConfig]
+
+
+def apply_media_transformation(
```
If these are wrapped in a class, then it should be pretty easy to make a consistent API with the `apply` method. 

WDYT?

juanitorduz replied:
yes!

wd60622 commented on pymc_marketing/mmm/media_transformation.py:
```diff
@@ -0,0 +1,278 @@
+            else (self.saturation, self.adstock)
+        )
+
+    def __call__(self, x, dim):
```
I use __call__ here instead of an `apply` method used in individual transformations. WYDT?
As mentioned below, we can have 

```python
media_data = ...
online_media_transformation = MediaTransformation(adstock=..., saturation=..., ...)
offline_media_transformation = MediaTransformation(adstock=..., saturation=..., ...)
media_configs = [MediaConfig(media_transformation=online_media_transformation), MediaConfig(media_transformation=offline_media_transformation)]
comb_media_transformation = CombinedMediaTransformation(media_configs=media_configs)

# Make all three work the same
offline_media_transformation(media_data, dims="media")
online_media_transformation(media_data, dims="media")
comb_media_transformation(media_data, dims="media")
```

This would have have many conveniences

juanitorduz replied:
Nice! 😎

wd60622 commented:
I've addressed all the feedback @juanitorduz  except for the integration with `MMM`. That would require a handful of adjustments and discussion. 

For instance, 
is this a new parameter for initialization? 

```python
media_transformation = MediaTransformation(...)
mmm = MMM(media_transformation=media_transformation, ...)

media_transformation = MediaConfigList(...)
mmm = MMM(media_transformation=media_transformation, ...)
```

That has some implications: 

- what do we do with current parameters `adstock`, `saturation`, `adstock_first`? Deprecate?
- How do we handle all of the model_config parameters? Should we store all of ModelTransformation / ModelConfigList there even though that is duplicated since there are stored in the individual transformations themselves
- Various plotting functions need to be rewritten in order to support multiple adstocks / saturations etc
- Budget optimizer will only conditionally work if single MediaTransformation

Definitely doable though

Do you want to do now or wait on this? Let me know how you imagine this being added in

juanitorduz reviewed (approved):
Yay! We should work on iterstions so let's merge this one and leave the MMM integration for another PR (can you please create an issue so that we keep track?)

wd60622 commented:
> Yay! We should work on iterstions so let's merge this one and leave the MMM integration for another PR (can you please create an issue so that we keep track?)

Yes, will make some issues now. 

Would you like me to increase the test coverage?

juanitorduz commented:
@wd60622 from my side it looks ok! If you wanna test edge cases go for it.

wd60622 commented:
> @wd60622 from my side it looks ok! If you wanna test edge cases go for it. 

I will create an issue for it
